From 02dc7b05c22af826e9506cdac3bfaa469514c039 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Thu, 18 Apr 2019 11:51:30 +0300
Subject: [PATCH 154/386] net: mvpp2: avoid infinity loops in cls flow find
 procedure

Brake classifier flow hash find loop if flow id is bigger
than classifier flow table size.
Infinity loop can cause CPU stall if something get wrong in
classifier configuration.

Change-Id: I3bb00e11dcffc853e3881f3d9acc68a18fc69295
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/7809
Reviewed-by: Marcin Wojtas <Marcin.Wojtas@cavium.com>
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c
index 226a9f433ae6..103c4a4fc8dc 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_cls.c
@@ -625,11 +625,13 @@ int mvpp2_cls_flow_hash_find(struct mvpp2_port *port,
 			     struct mvpp2_cls_flow_entry *fe,
 			     int *flow_index)
 {
-	int engine, is_last, flow_offset, port_bm, idx = 0;
+	int engine, flow_offset, port_bm, idx = 0, is_last = 0;
 
 	flow_offset = 0;
 	do {
 		idx = MVPP2_PORT_FLOW_INDEX(flow_offset, flow->flow_id);
+		if (idx >= MVPP2_CLS_FLOWS_TBL_SIZE)
+			break;
 		mvpp2_cls_flow_read(port->priv, idx, fe);
 		engine = mvpp2_cls_flow_eng_get(fe);
 		port_bm = mvpp2_cls_flow_port_get(fe);
-- 
2.17.1

